doc/sphinx/source/users-guide/datapar/forall.rst

```diff
@@ -0,0 +1,178 @@
+:title: Users Guide: **forall** loops
+
+**forall**-loops: data-parallel loops
+=====================================
+
+The **forall-loop** is Chapel's central concept for expressing data
+parallelism.  Forall-loops are similar to :ref:`coforall-loops
+<ug-coforall>` in that they both are parallel variants of Chapel's
+:ref:`for-loop <ug-forloops>`.  However, where a coforall-loop creates
+a concurrent task per iteration and a for-loop is executed serially by
+a single task, a forall-loop may use an arbitrary number of tasks to
+execute the loop.  As a result, it may execute serially using a single
+task, or it can use any number of tasks up to the number of iterations
+(or even beyond, though this is unusual).
+
+More specifically, the actual number of tasks used to execute a
+forall-loop is determined by the its iterand expressions.  The
+iterands invoke *parallel iterators*, either directly or via
+expressions that support default parallel iterator methods like
+ranges, domains, and arrays.  These iterators often select the number
+of tasks to use based on dynamic information such as the size of the
+loop and/or the number of available processors.  In addition to
+specifying the degree of concurrency to use, parallel iterators also
+determine where their tasks run and how the loop's iterations will be
+divided between them.  These choices are referred to as the
+*scheduling policy* for the loop.
+
+In practice, the default parallel iterators for many expressions
+create a number of tasks proportional to the amount of system
+parallelism that they are targeting.  For example, by default, a local
+forall loop will typically create a number of tasks equal to the
+number of available processor cores on its compute node.
+
+Like coforall-loops, the original task will not proceed beyond the
+loop until all tasks used to implement the loop have completed.
+
+As a simple motivating example, consider the following serial loop
+that increments each element in a million-element array *A*:
+
+.. literalinclude:: examples/users-guide/datapar/forall.chpl
+  :caption:
+  :language: chapel
+  :lines: 6-7,11-12
+
+Since the array is large and the loop's iterations are independent of
+one other, it should be advantageous to accelerate this computation by
+executing it in parallel.  A coforall-loop could be used to express
+the parallelism:
+
+.. literalinclude:: examples/users-guide/datapar/forall.chpl
+  :language: chapel
+  :lines: 16-17
+
+However, a coforall-loop would likely be overkill for this scenario
+since it would create one million tasks, each of which is simply
+responsible for incrementing a single element.  For most hardware
+platforms, this would generate more concurrency than the system could
+run in parallel while also creating very fine-grain tasks for which
+the cost of creating, scheduling, and destroying them would likely
+cost far more than the loop body itself.  Better would be to create a
```

Comment:
potential rewording:

For most hardware platforms, this would generate more concurrency than the system could run in parallel, and the overhead cost of creating, scheduling, and destroying each fine-grain task would be greater than the loop itself.